### PR TITLE
Draft: SC2-4-4 link-has-name

### PR DESCRIPTION
**Pull request by jkodu**
Draft rule submitted for issue - https://github.com/auto-wcag/auto-wcag/issues/101

### CODE REVIEW

**Comment by jkodu** [comment]
@annethyme @WilcoFiers @skotkjerra @Brynanders @kasperisager 
More approvals please.

File: _drafts/SC2-4-4-link-has-name.md [hunk 1]
```diff
@@ -35,51 +35,61 @@ Each target element has an [accessible name][] that is [non-empty][].
...
-```html
+### Failed
 
+```html
+<div>
+  <p>
+    Some detailed article, but the link to read more is placed with no relevance to the text, thereby the link has lost context.
+  </p>
+  <p>
+    <a href="readmore.html">Read More...</a>
```
Comment (annethyme) [inline _drafts/SC2-4-4-link-has-name.md]: This should be a pass in this rule. This rule only checks whether the link has a name, not whether it is appropriate.

File: _drafts/SC2-4-4-link-has-name.md [hunk 2]
```diff
@@ -35,51 +35,61 @@ Each target element has an [accessible name][] that is [non-empty][].
...
 ```html
-
+<table>
+   <tr> 
+       <td>Song: Ed Sheeran - Galway Girl.</td>
+   </tr>
+   <tr>
+       <td>
+        <a href="assets/download.html">
+          Download Now
```
Comment (annethyme) [inline _drafts/SC2-4-4-link-has-name.md]: This should be a pass in this rule. This rule only checks whether the link has a name, not whether it is appropriate.

File: _drafts/SC2-4-4-link-has-name.md [hunk 3]
```diff
@@ -35,51 +35,61 @@ Each target element has an [accessible name][] that is [non-empty][].
...
 
 ### Passed
 ```html
-<a href="http://www.w3.org/WAI">Web Accessibility Initiative (WAI)</a>
+<a href="http://www.w3.org/WAI"> Web Accessibility Initiative (WAI) </a>
 ```
 
 ```html
 <button role="link">Click me!</button>
 ```
```
Comment (annethyme) [inline _drafts/SC2-4-4-link-has-name.md]: I suggest adding passing test cases for at least these cases: 
- Link containing only an image (which I am pretty sure will be applicable to this rule, as it is written right now)
- Link containing both image and text (which I am pretty sure will be applicable to this rule, as it is written right now)
- Link labelled via e.g. aria-label, aria-labelledby, title etc.

File: _drafts/SC2-4-4-link-has-name.md [hunk 4]
```diff
@@ -35,51 +35,61 @@ Each target element has an [accessible name][] that is [non-empty][].
...
+   </tr>
+   <tr>
+       <td>
+        <a href="assets/download.html">
+          Download Now
+        </a>
+       </td>
+   </tr>
+ </table>
 ```
```
Comment (annethyme) [inline _drafts/SC2-4-4-link-has-name.md]: I suggest adding failing test cases for at least these cases: 
- Link containing only an image (which I am pretty sure will be applicable to this rule, as it is written right now)
- Link containing both image and text (which I am pretty sure will be applicable to this rule, as it is written right now)
- Link labelled via e.g. aria-label, aria-labelledby, title etc.
- Link that is not visible on page 
- Link that is not exposed to assistive technologies

File: pages/algorithms/visible-on-the-page.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+---
+title: Visible to the Page.
```
Comment (annethyme) [inline pages/algorithms/visible-on-the-page.md]: Visible to the Page --> Visible on the Page.

File: _drafts/SC2-4-4-link-has-name.md [hunk 6]
```diff
@@ -35,51 +35,61 @@ Each target element has an [accessible name][] that is [non-empty][].
...
 
 ### Inapplicable
 ```html
 <a href="http://www.w3.org/WAI" role="button">Web Accessibility Initiative (WAI)</a>
 ```
```
Comment (annethyme) [inline _drafts/SC2-4-4-link-has-name.md]: I suggest adding at least the following inapplicable test cases:
- Link that is neither visible on page or exposed to assistive technologies

File: _drafts/SC2-4-4-link-has-name.md [hunk 7]
```diff
@@ -1,5 +1,5 @@
 ---
```
Comment (annethyme) [inline _drafts/SC2-4-4-link-has-name.md]: In the Applicability I suggest making two changes:
- are --> is
- having "semantic role" link to a definition and delete the Note.

"The rule applies to any HTML or SVG element with the semantic role of link that are ..." + **Note**:  
--> "The rule applies to any HTML or SVG element with the [semantic role][] of link that is ..." (+ no note)

File: _drafts/SC2-4-4-link-has-name.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+There are no major accessibility support issues known for this rule.
+
+## Background
+
+- [https://www.w3.org/WAI/WCAG20/quickref/?showtechniques=244#navigation-mechanisms-refs](https://www.w3.org/WAI/WCAG20/quickref/?showtechniques=244#navigation-mechanisms-refs)
+- [ARIA7: Using aria-labelledby for link purpose](https://www.w3.org/TR/WCAG20-TECHS/ARIA7.html)
+- [ARIA8: Using aria-label for link purpose](https://www.w3.org/TR/WCAG20-TECHS/ARIA8.html)
+- [F89: Failure of Success Criteria 2.4.4, 2.4.9 and 4.1.2 due to using null alt on an image where the image is the only content in a link](http://www.w3.org/TR/WCAG20-TECHS/F89.html)
+- The WCAG 2.0 Techniques already contain examples and code snippets to illustrate which content passes or fails the test. Whenever possible auto-wcag refers to those. Another source for test cases is the W3C Before and After Demonstration.
```
Comment (annethyme) [inline _drafts/SC2-4-4-link-has-name.md]: Isn't this a piece from the template that has snuck into the rule?

File: _drafts/SC2-4-4-link-has-name.md (hunk 4 above)
Comment (annethyme) [reply _drafts/SC2-4-4-link-has-name.md]: This one doesn't make sense, so it's not incorporated: Link containing both image and text (which I am pretty sure will be applicable to this rule, as it is written right now)

**Pull request merged**